sovrin_node/persistence/idr_cache.py

```diff
@@ -91,18 +94,28 @@ def close(self):
         self._keyValueStorage.close()
 
     def currentBatchCreated(self, stateRoot):
-        self.unCommitted[stateRoot] = OrderedDict(self.currentBatchOps)
+        self.unCommitted.append((stateRoot, OrderedDict(self.currentBatchOps)))
         self.currentBatchOps = []
 
-    def batchRejected(self, stateRoot=None):
-        if stateRoot:
-            self.unCommitted[stateRoot] = OrderedDict(self.currentBatchOps)
-        else:
-            self.currentBatchOps = []
+    def batchRejected(self):
+        # Batches are always rejected from end of `self.unCommitted`
+        self.currentBatchOps = []
+        self.unCommitted = self.unCommitted[:-1]
 
     def onBatchCommitted(self, stateRoot):
-        self._keyValueStorage.setBatch([(idr, val) for idr, val in self.unCommitted[stateRoot].items()])
-        self.unCommitted.pop(stateRoot)
+        # Commit an already created batch
+        if self.unCommitted:
+            assert self.unCommitted[0][0] == stateRoot, 'The first created batch has ' \
+                                                     'not been committed or ' \
+                                                     'reverted and yet another ' \
+                                                     'batch is trying to be ' \
+                                                     'committed'
+            self._keyValueStorage.setBatch([(idr, val) for idr, val in
+                                            self.unCommitted[0][1].items()])
+            self.unCommitted = self.unCommitted[:-1]
```

Comment:
Shouldn't it be `self.unCommitted = self.unCommitted[1:]` so that we remove the first (processed item), not the last one.